### ISSUE TITLE
'owner-id' for DescribeImages is not yet supported

### ISSUE THREAD

**Issue by snimark** [spulec/moto#1200]
i am trying to test out a routine to manage our AMIs. the process involves filtering the AMIs by owner (aws account id) of the images to pull back a smaller set from aws. turns out, `owner-id` is not yet supported.

### moto, boto3, botocore
i installed it using `sudo pip3 install -r requirements.txt` and moto is in that file.

Name: moto
Version: 1.0.1
Summary: A library that allows your python tests to easily mock out the boto library
Home-page: https://github.com/spulec/moto
Author: Steve Pulec
Author-email: spulec@gmail.com
License: Apache
Location: /usr/lib/python3.5/site-packages
Requires: requests, python-dateutil, cookies, mock, boto3, six, werkzeug, pyaml, xmltodict, Jinja2, pytz, dicttoxml, boto

Name: boto3
Version: 1.4.4
Summary: The AWS SDK for Python
Home-page: https://github.com/boto/boto3
Author: Amazon Web Services
Author-email: UNKNOWN
License: Apache License 2.0
Location: /usr/lib/python3.5/site-packages
Requires: jmespath, s3transfer, botocore

Name: botocore
Version: 1.5.95
Summary: Low-level, data-driven core of boto 3.
Home-page: https://github.com/boto/botocore
Author: Amazon Web Services
Author-email: UNKNOWN
License: Apache License 2.0
Location: /usr/lib/python3.5/site-packages
Requires: python-dateutil, jmespath, docutils

### stack trace from pytest
```
$ pytest -v test/test_manage_amis.py 
================================================================ test session starts =================================================================
platform linux -- Python 3.5.3, pytest-3.2.2, py-1.4.34, pluggy-0.4.0 -- /usr/bin/python3
cachedir: .cache
rootdir: /home/mock/projects/engineering-puppet, inifile:
collected 1 item                                                                                                                                      

test/test_manage_amis.py::TestManageAmis::test_get_images_generated_by_deployment_tools FAILED

====================================================================== FAILURES ======================================================================
____________________________________________ TestManageAmis.test_get_images_generated_by_deployment_tools ____________________________________________

self = <test_manage_amis.TestManageAmis testMethod=test_get_images_generated_by_deployment_tools>

    @mock_ec2
    @mock_sts
    def test_get_images_generated_by_deployment_tools(self):
        # setup
        resource = boto3.resource('ec2', region_name='us-east-1')
        client = boto3.client('ec2', region_name='us-east-1')
        session = manage_amis.get_session('111133335555', region='us-east-1')
    
        all_images = list()
        ami_tags = {'Key': 'tools_info',
                    'Value': str({'user': 'jenkins', 'aws-ami': 'ami-ffff9999'})}
    
        # generate tools images
        tools_ami_ids = ['1234abcd', 'abcd1234', '1a2b3c4d', 'a1b2c3d4', '12ab34cd', 'dc43ba21']
        for ami_id in tools_ami_ids:
            response = client.run_instances(ImageId="ami-{}".format(ami_id), MinCount=1, MaxCount=2)
            instance_id = response['Instances'][0]['InstanceId']
            image = client.create_image(InstanceId=instance_id, Name="test-{}".format(ami_id))
            client.create_tags(Resources=[image['ImageId']], Tags=[ami_tags])
            all_images.append(image['ImageId'])
            client.terminate_instances(InstanceIds=[instance_id])
            client.run_instances(ImageId=image['ImageId'], MinCount=1, MaxCount=2)
    
        # generate non tools images
        non_tools_ami_ids = ['5678efab', 'efab5678', '5a6b7c8d']
        for ami_id in non_tools_ami_ids:
            response = client.run_instances(ImageId="ami-{}".format(ami_id), MinCount=1, MaxCount=2)
            instance_id = response['Instances'][0]['InstanceId']
            image = client.create_image(InstanceId=instance_id, Name="test-{}".format(ami_id))
            all_images.append(image['ImageId'])
            client.terminate_instances(InstanceIds=[instance_id])
            client.run_instances(ImageId=image['ImageId'], MinCount=1, MaxCount=2)
    
    
        # test the function
>       amis = manage_amis.get_images_generated_by_deployment_tools(session, '111133335555')

test/test_manage_amis.py:77: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
manage-amis.py:186: in get_images_generated_by_deployment_tools
    images = ec2.describe_images(Filters=image_filter)['Images']
/usr/lib/python3.5/site-packages/botocore/client.py:310: in _api_call
    return self._make_api_call(operation_name, kwargs)
/usr/lib/python3.5/site-packages/botocore/client.py:586: in _make_api_call
    operation_model, request_dict)
/usr/lib/python3.5/site-packages/botocore/endpoint.py:141: in make_request
    return self._send_request(request_dict, operation_model)
/usr/lib/python3.5/site-packages/botocore/endpoint.py:170: in _send_request
    success_response, exception):
/usr/lib/python3.5/site-packages/botocore/endpoint.py:249: in _needs_retry
    caught_exception=caught_exception, request_dict=request_dict)
/usr/lib/python3.5/site-packages/botocore/hooks.py:227: in emit
    return self._emit(event_name, kwargs)
/usr/lib/python3.5/site-packages/botocore/hooks.py:210: in _emit
    response = handler(**kwargs)
/usr/lib/python3.5/site-packages/botocore/retryhandler.py:183: in __call__
    if self._checker(attempts, response, caught_exception):
/usr/lib/python3.5/site-packages/botocore/retryhandler.py:251: in __call__
    caught_exception)
/usr/lib/python3.5/site-packages/botocore/retryhandler.py:269: in _should_retry
    return self._checker(attempt_number, response, caught_exception)
/usr/lib/python3.5/site-packages/botocore/retryhandler.py:317: in __call__
    caught_exception)
/usr/lib/python3.5/site-packages/botocore/retryhandler.py:223: in __call__
    attempt_number, caught_exception)
/usr/lib/python3.5/site-packages/botocore/retryhandler.py:359: in _check_caught_exception
    raise caught_exception
/usr/lib/python3.5/site-packages/botocore/endpoint.py:204: in _get_response
    proxies=self.proxies, timeout=self.timeout)
/usr/lib/python3.5/site-packages/botocore/vendored/requests/sessions.py:573: in send
    r = adapter.send(request, **kwargs)
/usr/lib/python3.5/site-packages/moto/packages/responses/responses.py:302: in unbound_on_send
    return self._on_request(adapter, request, *a, **kwargs)
/usr/lib/python3.5/site-packages/moto/packages/responses/responses.py:255: in _on_request
    status, r_headers, body = match['callback'](request)
/usr/lib/python3.5/site-packages/moto/core/utils.py:164: in __call__
    result = self.callback(request, request.url, request.headers)
/usr/lib/python3.5/site-packages/moto/core/responses.py:111: in dispatch
    return cls()._dispatch(*args, **kwargs)
/usr/lib/python3.5/site-packages/moto/core/responses.py:179: in _dispatch
    return self.call_action()
/usr/lib/python3.5/site-packages/moto/core/responses.py:196: in call_action
    response = method()
/usr/lib/python3.5/site-packages/moto/ec2/responses/amis.py:48: in describe_images
    ami_ids=ami_ids, filters=filters, exec_users=exec_users)
/usr/lib/python3.5/site-packages/moto/ec2/models.py:1068: in describe_images
    return generic_filter(filters, images)
/usr/lib/python3.5/site-packages/moto/ec2/utils.py:493: in generic_filter
    objects = [obj for obj in objects if is_filter_matching(
/usr/lib/python3.5/site-packages/moto/ec2/utils.py:494: in <listcomp>
    obj, _filter, _filter_value)]
/usr/lib/python3.5/site-packages/moto/ec2/utils.py:471: in is_filter_matching
    value = obj.get_filter_value(filter)
/usr/lib/python3.5/site-packages/moto/ec2/models.py:1026: in get_filter_value
    "The filter '{0}' for DescribeImages".format(filter_name))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <moto.ec2.models.EC2Backend object at 0x7f920f63df98>, blurb = "The filter 'owner-id' for DescribeImages"

    def raise_not_implemented_error(self, blurb):
        msg = "{0} has not been implemented in Moto yet." \
              " Feel free to open an issue at" \
              " https://github.com/spulec/moto/issues".format(blurb)
>       raise NotImplementedError(msg)
E       NotImplementedError: The filter 'owner-id' for DescribeImages has not been implemented in Moto yet. Feel free to open an issue at https://github.com/spulec/moto/issues

/usr/lib/python3.5/site-packages/moto/ec2/models.py:3596: NotImplementedError
---------------------------------------------------------------- Captured stderr call ----------------------------------------------------------------
2017-09-26 15:05:32,581 [INFO    ] : root : get_session : 118 : Getting session token...
2017-09-26 15:05:32,954 [INFO    ] : root : get_images_generated_by_deployment_tools : 185 : Fetching AMI objects for account id 111133335555...
```
### function call
the call where the tag is being specified is in the `manage-ami.py` file. the routing from that below:
```
def get_images_generated_by_deployment_tools(session, account_id):
    """
    get all the images generated by our deployment tools.
    param session: boto3 session ot establish connection to account
     type session: boto3.session.Session class
    param accound_id: id of the account to filter the images down to ours
     type account_id: string
    returns: list of ...
    """
    # get an ec2 session client
    LOG.debug("Creating EC2 session...")
    ec2 = session.client('ec2', region_name='us-east-1')

    # this process is similar to the tag_images fetch, but the filter is different here.
    # we want to narrow down this list as much as possible. and not filtering by account 
    # id will return ALL AMIs, over 85,000 at the time of this coding.
    LOG.debug("""Set filter to look for owner-id = "{}" and tag-key = "tools_info" """.format(account_id))
    image_filter = [{'Name': 'owner-id', 'Values': [account_id]},
                    {'Name': 'tag-key', 'Values': ['tools_info']}]

    # we only care about the images
    LOG.info("Fetching AMI objects for account id {}...".format(account_id))
    images = ec2.describe_images(Filters=image_filter)['Images']

    # we only care about the image ids of the cared-about images from above
    LOG.debug("Getting AMI ids from the list of images...")
    image_ids = [v for i in images for k,v in i.items() if k == "ImageId"]

    return image_ids
```

**Comment by snimark** [comment]
and honestly, i've thought about adjusting the code and submitting a patch, but i would need to probably consult with you guys on the how of it. AMIs have an owner id value of the aws account id. when it's not specified, would you just assign a dummy value or leave it blank (None)?

**Comment by JackDanger** [comment]
@snimark Your intuition is as good as anybody's on how to implement it. I'd have a default account_id assigned to a couple of built-in AMIs (maybe real ones that are commonly used in EC2) and then use the credentials of inbound requests to assign the owner to any images created through the API. You can get the account id out of the 'Authorization' header in the request when Moto is running in server mode.

**Comment by snimark** [comment]
is the idea of submissions for changes via pull requests welcomed for this project? if so, do you have a policy on how to branch (naming and otherwise) i should follow to do that?

**Comment by JackDanger** [comment]
> is the idea of submissions for changes via pull requests welcomed for this project?

Extremely welcome! Bordering on mandatory :p 

> if so, do you have a policy on how to branch (naming and otherwise) i should follow to do that?

No policy on that other than following the [Code of Conduct](https://github.com/spulec/moto/blob/master/CODE_OF_CONDUCT.md). Just make sure any code you add is covered by reasonable tests and ask people to jump in and help whenever you get stuck. We're a friendly bunch here at Moto :)

**Comment by snimark** [comment]
i have a couple of questions:
1. what is the `ec2_backend` value passed in to the `class AMI()` call?
2. i am thinking of making this an optional argument passed in for object instantiation. with instance argument value being in, would it have an `owner-id` value to be assigned? i would think a `source_ami` _could_ have an `owner-id` value, depending on if it was created or not:

this code is currently the class for the AMI (i removed the quoted section after the `elif source_ami:` line):

```class Ami(TaggedEC2Resource):
    def __init__(self, ec2_backend, ami_id, instance=None, source_ami=None,
                 owner_id=None, name=None, description=None):
        self.ec2_backend = ec2_backend
        self.id = ami_id
        self.state = "available"
        self.name = name
        self.description = description
        self.virtualization_type = None
        self.architecture = None
        self.kernel_id = None
        self.platform = None
        self.owner_id = owner_id
        self.creation_date = utc_date_and_time()

        if instance:
            self.instance = instance
            self.instance_id = instance.id
            self.virtualization_type = instance.virtualization_type
            self.architecture = instance.architecture
            self.kernel_id = instance.kernel
            self.platform = instance.platform

        elif source_ami:
            self.virtualization_type = source_ami.virtualization_type
            self.architecture = source_ami.architecture
            self.kernel_id = source_ami.kernel_id
            self.platform = source_ami.platform
            if not name:
                self.name = source_ami.name
            if not description:
                self.description = source_ami.description
            if source_ami.owner_id:
                self.owner_id = source_ami.owner_id
...
```
i know i would have to do some work with the `get_filter_value()` call too, but i think that might be easier knowing the answers above.

**Comment by snimark** [comment]
i figured out what ec2_backend is. region, right?

i am modifying also the `Instance()` class to also include the `owner-id` property. and i found the unittests to also write to as well.

**Comment by snimark** [comment]
hmm...are we sure the images (AMIs) being assigned but with values of `123456789012`. i didn't see this, so i tried to track it down.

**Comment by snimark** [comment]
ok, well, i'm going to need to chat with someone to work through this option. although i found the OwnerId/owner-id value is set by default for instances to be `123456789012`, but i also found it does not translate to the AMI classes. all attempts to set the value from the front or add the field from the back have not worked.

i joined gitter to chat, but the channel seems very quiet these days. i will continue to check back, but i think i am going to have to test without moto for now.

**Comment by JackDanger** [comment]
@toshitanian or @terrycain do either of you have opinions here?

**Comment by snimark** [comment]
i can create a branch an submit it with the changes i've been attempting. i can try to add notes where i am not seeing it work.

**Comment by terrycain** [comment]
Yeah I think what we need would be, a default set of ami's present, though they do change so we'd need to manage that somehow. We also need to document how you can just import the moto backend and prod all the structures. @snimark make a PR and note down as much as you can plus a test case where it doesnt work and should.

**Comment by snimark** [comment]
i see to be having permission problems creating a branch locally and pushing to github. my permissions are in question as far as the push goes. (i can get to github with my current key.)

**Comment by terrycain** [comment]
Have you created a fork in your own account and are pushing to that?

**Comment by snimark** [comment]
no, nothing in my github account. i just cloned the repo locally and made the changes. then was trying to push the branch back to the repo.

```$ git push -u origin mock/adding-owner-id-to-amis
ERROR: Permission to spulec/moto.git denied to snimark.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```
yet...
```$ ssh -T git@github.com
Hi snimark! You've successfully authenticated, but GitHub does not provide shell access.
```

**Comment by terrycain** [comment]
Nah, you need to click on fork at the top of the page to clone it to your own account, then you'll be allowed to push to that, then you can create a pull request from the clone in your account to this one

**Comment by snimark** [comment]
ah, ok.

i have tried this, gotten to the pull request, but one of the files is showing changes i didn't make, so i will need to do this again to make sure the files are correct. i need to head out right now, but i'll try to submit later.

**Comment by snimark** [comment]
ok, finally got my branch created on the forked repo and created a pull request. i have not created a unit test for it, though. i have another script i can make available which i have been using to simulate the unit test i need for my code.

**Comment by terrycain** [comment]
Wait so do the changes work or not? And yeah the script will be helpful, throw it in a gist would probably be easiest.

**Comment by snimark** [comment]
they weren't, but i think i had success in one case of testing after fast forwarding my changes to a few updates to the master branch. i need to test these again to evaluate what is going on. i also notices that when i had submitted by branch that a few mods were missing. i would need to submit a new pull request when i get this more sorted out.

**Comment by terrycain** [comment]
Sure once your at a point where either its working or you have no clue why its not, give me a mention and I'll have as look over it.

**Comment by snimark** [comment]
i apologize for the severely delayed response. i ended up taking a different approach for the time and then had other matters pull my focus.

to answer your question in the past, no, i didn't work as expected at the time. i needed to go back and start looking to other parts of the moto code to make sure i was accounting for my changes beyond the modules on which i had made suggestions.

i realize that everyone has moved on and there has been progress made. i will revisit how things have advanced. it's most likely someone has done better work and rendered my work unnecessary at this stage. i hope to do better at my community involvement going forward.
